### PR TITLE
Setup travis and gemfile to test differing versions of rspec

### PR DESCRIPTION
**Pull request by JonRowe**
We should test compatibility across a range of rspec versions

### CODE REVIEW

File: .travis.yml [hunk 1]
```diff
@@ -18,4 +18,14 @@ matrix:
...
+    - rvm: 2.2
+      env: RSPEC='3-0-maintenance'
+    - rvm: 2.2
+      env: RSPEC='3-1-maintenance'
+    - rvm: 2.2
+      env: RSPEC='3-2-maintenance'
+    - rvm: 2.2
+      env: RSPEC='3-3-maintenance'
+    - rvm: 2.2
+      env: RSPEC='3-4-maintenance'
```
Comment (myronmarston) [inline .travis.yml]: I think `RSPEC_BRANCH` would be a clearer name than `RSPEC` for this env var.

File: script/functions.sh [hunk 2]
```diff
@@ -59,6 +59,46 @@ function rspec_support_compatible {
...
+function run_other_repo_specs {
+  fold "rspec-core specs" run_spec_suite_for "rspec-core"
+  fold "rspec-expectations specs" run_spec_suite_for "rspec-expectations"
+  fold "rspec-mocks specs" run_spec_suite_for "rspec-mocks"
+  fold "rspec-rails specs" run_spec_suite_for "rspec-rails"
+
+  if rspec_support_compatible; then
+    fold "rspec-support specs" run_spec_suite_for "rspec-support"
+  fi
+}
```
Comment (myronmarston) [inline script/functions.sh]: I don't think we need to run the specs for all the rspec repos as part of builds for this repo.

File: .travis.yml (hunk 1 above)
Comment (JonRowe) [inline .travis.yml]: Yeah I'm not entirely sure this approach will even work at the moment, I just want to see what travis does :)

File: script/functions.sh (hunk 2 above)
Comment (JonRowe) [inline script/functions.sh]: It's just a sample spec base for now

**Comment by JonRowe** [comment]
Ready for review, although it doesn't catch #18 yet, going to write some more specific specs but this sets the build up so that can be a follow up PR

**Comment by JonRowe** [comment]
Care to review @myronmarston?

File: script/functions.sh [hunk 3]
```diff
@@ -82,7 +84,32 @@ function run_specs_and_record_done {
...
+    unset BUNDLE_GEMFILE
+    bundle_install_flags=`cat ../.travis.yml | grep bundler_args | tr -d '"' | grep -o " .*"`
+    travis_retry bundle install $bundle_install_flags
+    cp ../.rspec .rspec
+    set +e
+    bin/rspec spec ../spec --format NyanCatFormatter --out $SPECS_HAVE_RUN_FILE
+    local status=$?
+    set -e
+    popd
+    if test $status = 42; then
```
Comment (myronmarston) [inline script/functions.sh]: Why 42?

File: script/functions.sh [hunk 4]
```diff
@@ -82,7 +84,32 @@ function run_specs_and_record_done {
     rspec_bin=script/rspec_with_simplecov
   fi;
 
-  $rspec_bin spec --backtrace --format progress --profile --format progress --out $SPECS_HAVE_RUN_FILE
+  $rspec_bin spec --backtrace --profile --format progress --out $SPECS_HAVE_RUN_FILE
```
Comment (myronmarston) [inline script/functions.sh]: Why the change here?

File: script/functions.sh [hunk 5]
```diff
@@ -82,7 +84,32 @@ function run_specs_and_record_done {
...
+    pushd ./smoke_specs
+    echo
+    echo "Running smoke specs for $RSPEC_VERSION"
+    echo
+    unset BUNDLE_GEMFILE
+    bundle_install_flags=`cat ../.travis.yml | grep bundler_args | tr -d '"' | grep -o " .*"`
+    travis_retry bundle install $bundle_install_flags
+    cp ../.rspec .rspec
+    set +e
+    bin/rspec spec ../spec --format NyanCatFormatter --out $SPECS_HAVE_RUN_FILE
```
Comment (myronmarston) [inline script/functions.sh]: In the travis output we're able to see that it ran this line but we don't see any of the output from it because it's only going into the file.  That seems unfortunate.  Any way to avoid that?

File: script/functions.sh (hunk 3 above)
Comment (JonRowe) [inline script/functions.sh]: To differentiate between ruby crashes and a correctly failing test suite.

Comment (myronmarston) [inline script/functions.sh]: But RSpec usually exits with a status of 1 when there's a failure, so where did 42 come from?

File: script/functions.sh (hunk 4 above)
Comment (JonRowe) [inline script/functions.sh]: Duplicate arguments, only noticed when tinkering with the line

File: script/functions.sh (hunk 3 above)
Comment (myronmarston) [inline script/functions.sh]: Nevermind, I see where you configure it below now.  I think a comment here explaining why 42 would help.

File: script/functions.sh (hunk 5 above)
Comment (JonRowe) [inline script/functions.sh]: Could add `cat $SPECS_HAVE_RUN_FILE`?

File: script/functions.sh (hunk 4 above)
Comment (myronmarston) [inline script/functions.sh]: The duplication is intentional -- the first `--format progress` sets the formatter for stdout and then `--format progress --out $SPECS_HAVE_RUN_FILE` uses a second instance of the progress formatter that outputs to `$SPECS_HAVE_RUN_FILE`.

File: script/functions.sh (hunk 5 above)
Comment (myronmarston) [inline script/functions.sh]: that would probably work, but why not just do:

``` bash
bin/rspec spec ../spec --format NyanCatFormatter --format NyanCatFormatter --out $SPECS_HAVE_RUN_FILE
```

...which would use two instances of NyanCatFormatter, one to stdout and one to `$SPECS_HAVE_RUN_FILE`?  Or you could use `tee`.

File: script/functions.sh (hunk 4 above)
Comment (JonRowe) [inline script/functions.sh]: Oh ok :)

File: script/functions.sh (hunk 3 above)
Comment (JonRowe) [inline script/functions.sh]: :+1:

**Comment by JonRowe** [comment]
Updated with your comments in mind @myronmarston

**Comment by myronmarston** [comment]
The travis build is failing.

**Comment by JonRowe** [comment]
> The travis build is failing.

Yep, seems bundler is crapping itself, unrelated to these changes.

**Comment by myronmarston** [comment]
JRuby doesn't seem to like this, but it LGTM otherwise.

**Comment by JonRowe** [comment]
We don't run JRuby 2.0 elsewhere as it's not stable, so lets not do so here either, it can't install / update bundler it would seem, my intent is to merge when green.
